From 70e8b42963fcd43478add09ac737c2aaacd205b4 Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@nxp.com>
Date: Fri, 9 Mar 2018 12:41:42 +0200
Subject: [PATCH 3472/5242] MLK-17580: ASoC: fsl: sai: check for pinctrl
 status

commit  fc7835533d570a65e336b3eb03777cceb2625ddc from
https://source.codeaurora.org/external/imx/linux-imx.git

For some cases (like AMIX) pinctrl may be null - this
breaks SAI functionality. Enforce pinctrl null pointer
checking prior calling any function which involves
pins state changes.

Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_sai.c |   21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/sound/soc/fsl/fsl_sai.c b/sound/soc/fsl/fsl_sai.c
index 17194355..21e8bb9 100644
--- a/sound/soc/fsl/fsl_sai.c
+++ b/sound/soc/fsl/fsl_sai.c
@@ -528,14 +528,17 @@ static int fsl_sai_hw_params(struct snd_pcm_substream *substream,
 
 	pins = DIV_ROUND_UP(channels, slots);
 	sai->is_dsd = fsl_is_dsd(params);
-	sai->pins_state = fsl_get_pins_state(sai->pinctrl, params);
 
-	if (!IS_ERR_OR_NULL(sai->pins_state)) {
-		ret = pinctrl_select_state(sai->pinctrl, sai->pins_state);
-		if (ret) {
-			dev_err(cpu_dai->dev,
-				"failed to set proper pins state: %d\n", ret);
-			return ret;
+	if (!IS_ERR_OR_NULL(sai->pinctrl)) {
+		sai->pins_state = fsl_get_pins_state(sai->pinctrl, params);
+
+		if (!IS_ERR_OR_NULL(sai->pins_state)) {
+			ret = pinctrl_select_state(sai->pinctrl, sai->pins_state);
+			if (ret) {
+				dev_err(cpu_dai->dev,
+					"failed to set proper pins state: %d\n", ret);
+				return ret;
+			}
 		}
 	}
 
@@ -909,7 +912,7 @@ static int fsl_sai_dai_resume(struct snd_soc_dai *cpu_dai)
 	struct fsl_sai *sai = snd_soc_dai_get_drvdata(cpu_dai);
 	int ret;
 
-	if (!IS_ERR_OR_NULL(sai->pins_state)) {
+	if (!IS_ERR_OR_NULL(sai->pinctrl) && !IS_ERR_OR_NULL(sai->pins_state)) {
 		ret = pinctrl_select_state(sai->pinctrl, sai->pins_state);
 		if (ret) {
 			dev_err(cpu_dai->dev,
@@ -1335,7 +1338,7 @@ static int fsl_sai_probe(struct platform_device *pdev)
 	sai->dma_params_rx.maxburst = FSL_SAI_MAXBURST_RX;
 	sai->dma_params_tx.maxburst = FSL_SAI_MAXBURST_TX;
 
-	sai->pinctrl  = devm_pinctrl_get(&pdev->dev);
+	sai->pinctrl = devm_pinctrl_get(&pdev->dev);
 
 	platform_set_drvdata(pdev, sai);
 
-- 
1.7.9.5

